presto-hive/src/main/java/com/facebook/presto/hive/S3SelectPushdown.java

```diff
@@ -0,0 +1,193 @@
+            "s3://", "s3a://", "s3n://");
+    private static final Set<String> SUPPORTED_SERDES = ImmutableSet.of(
+            "org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe");
+    private static final Set<String> SUPPORTED_INPUT_FORMATS = ImmutableSet.of(
+            "org.apache.hadoop.mapred.TextInputFormat");
+    private static final Set<String> SUPPORTED_COLUMN_TYPES = ImmutableSet.of(
+            "boolean",
+            "int",
+            "tinyint",
+            "smallint",
+            "bigint",
+            "string",
+            "float",
+            "double",
+            "decimal",
+            "date",
+            "timestamp");
+
+    private final HiveClientConfig hiveClientConfig;
+
+    S3SelectPushdown(HiveClientConfig hiveClientConfig)
+    {
+        this.hiveClientConfig = requireNonNull(hiveClientConfig);
+    }
+
+    private static boolean isSerdeSupported(Properties schema)
+    {
+        String serdeName = getDeserializerClassName(schema);
+        return SUPPORTED_SERDES.contains(serdeName);
+    }
+
+    private static boolean isInputFormatSupported(Properties schema)
+    {
+        String inputFormat = getInputFormatName(schema);
+        return SUPPORTED_INPUT_FORMATS.contains(inputFormat);
+    }
+
+    public static boolean isCompressionCodecSupported(InputFormat inputFormat, Path path)
+    {
+        if (inputFormat instanceof TextInputFormat) {
+            try {
+                Optional<CompressionCodec> compressionCodec = getCompressionCodec((TextInputFormat) inputFormat, path);
+                if (!compressionCodec.isPresent() ||
+                        stream(S3SelectSupportedCompressionCodecs.values())
+                                .anyMatch(codec -> codec.getCodecClass().isInstance(compressionCodec.get()))) {
+                    return true;
+                }
+            }
+            catch (IllegalAccessException | NoSuchFieldException e) {
+                LOG.error(e, "Failed to find compressionCodec for inputFormat: %s", inputFormat.getClass().getName());
+                return false;
+            }
+        }
+        return false;
+    }
+
+    private static boolean areColumnTypesSupported(List<Column> columns)
+    {
+        if (columns == null || columns.isEmpty()) {
+            return false;
+        }
+
+        for (Column column : columns) {
+            String type = column.getType().getHiveTypeName().toString();
+            if (type.startsWith(StandardTypes.DECIMAL)) {
+                // skip precision and scale when check decimal type
+                type = StandardTypes.DECIMAL;
+            }
+            if (!SUPPORTED_COLUMN_TYPES.contains(type.toLowerCase(Locale.ENGLISH))) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    private boolean isS3SelectEnabled(ConnectorSession session)
+    {
+        if (!hiveClientConfig.isS3SelectPushdownEnabled()) {
```

Comment:
We shouldn't check the config here, since config is used as the default session property value. Treating it as an override is confusing and different than everything else. It also precludes disabling it by default but enabling it for specific queries with a session property.

Reply:
After removing this check, we shouldn't need this stateful class anymore.

Reply:
> We shouldn't check the config here, since config is used as the default session property value. Treating it as an override is confusing and different than everything else.

Could you please elaborate on this. My current understanding is that we need a member variable in S3SelectPushdown to extract the boolean value from hiveClientConfig. However, it is not clear to me as to how that would allow us to enable/disable the pushdown by setting session property?

Reply:
The pattern for session properties is that the config sets the default value for the session property. The code that uses the feature only retrieves the session property. Let's assume that we set the config to false by default. Here's how it would work:

* User doesn't set config or session property: feature disabled
* User sets config to true: feature enabled (because session default is true)
* User sets config to true and sets session property to false: feature disabled (session default overridden)
* User doesn't set config but sets session property: feature enabled

This allows selectively enabling the feature for specific queries, or enabling for everything but disabling for specific queries.

Reply:
> Are you referring to hiveClientConfig, as the stateful class?

I was referring to this class, `S3SelectPushdown`

Reply:
Thanks for the elaboration and clarification. Fixed.